pyutilib/misc/timing.py

```diff
@@ -160,3 +160,305 @@ def start(self):
+
+    """
+    def __init__(self):
+        self.stack = list()
+        self.timers = dict()
+
+    def _get_timer(self, identifier, should_exist=False):
+        """
+
+        This method gets the timer associated with the current state
+        of self.stack and the specified identifier.
+
+        Parameters
+        ----------
+        identifier: str
+            The name of the timer
+        should_exist: bool
+            The should_exist is True, and the timer does not already
+            exist, an error will be raised. If should_exist is False, and
+            the timer does not already exist, a new timer will be made.
+        
+        Returns
+        -------
+        timer: _HierarchicalHelper
+
+        """
+        parent = self._get_timer_from_stack(self.stack)
+        if identifier in parent.timers:
+            return parent.timers[identifier]
+        else:
+            if should_exist:
+                raise RuntimeError('Could not find timer {0}'.format('.'.join(self.stack + [identifier])))
+            parent.timers[identifier] = _HierarchicalHelper()
+            return parent.timers[identifier]
+
+    def start_increment(self, identifier):
+        """
+        Start incrementing the timer identified with identifier
+
+        Parameters
+        ----------
+        identifier: str
+            The name of the timer
+        """
+        timer = self._get_timer(identifier)
+        timer.start_increment()
+        self.stack.append(identifier)
+
+    def stop_increment(self, identifier):
+        """
+        Stop incrementing the timer identified with identifier
+
+        Parameters
+        ----------
+        identifier: str
+            The name of the timer
+        """
+        if identifier != self.stack[-1]:
+            raise ValueError(str(identifier) + ' is not the currently active timer. The only timer that can currently be stopped is ' + '.'.join(self.stack))
+        self.stack.pop()
+        timer = self._get_timer(identifier, should_exist=True)
+        timer.stop_increment()
+
+    def __str__(self):
+        max_name_len = 12
+        for name in self.timers.keys():
+            if len(name) > max_name_len:
+                max_name_len = len(name)
+        name_formatter = '{name:<' + str(max_name_len) + '}'
+        s = (name_formatter +
+             '{total_time:>15}'
+             '{num_calls:>15}'
+             '{time_per_call:>15}\n').format(name='Identifier',
+                                             total_time='Time (s)',
+                                             num_calls='# Calls',
+                                             time_per_call='Per Call (s)')
+        s += '-' * (max_name_len + 45) + '\n'
+        for name, timer in self.timers.items():
+            s += name_formatter.format(name=name)
+            s += '{0:>15.2e}'.format(timer.total_time)
+            s += '{0:>15d}'.format(timer.n_calls)
+            s += '{0:>15.2e}\n'.format(timer.total_time/timer.n_calls)
```

Comment:
Can `n_calls` ever be 0?

Reply:
No, n_calls can never be 0. The timer would just not exist in that case.